File: thesis_visualizations.R
# Visualizations — Ferroptosis Hub Genes & Apoptosis in Ovarian Cancer
# MSc Bioinformatics Dissertation, Pondicherry University (2024)
# Author: Shruti Banerjee
#
# Run: Rscript visualizations.R
# Output: PNG files saved to charts/ folder
# Requirements: install.packages(c("ggplot2", "dplyr"))

library(ggplot2)

dir.create("charts", showWarnings = FALSE)

source_note <- "Source: Banerjee S. (2024). MSc Dissertation, Pondicherry University."
PURPLE <- "#534AB7"
TEAL   <- "#1D9E75"
CORAL  <- "#D85A30"
AMBER  <- "#BA7517"
GRAY   <- "#888780"

theme_thesis <- theme_minimal(base_size = 12) +
  theme(
    plot.title    = element_text(size = 12, face = "bold", margin = margin(b = 8)),
    plot.subtitle = element_text(size = 10, color = "gray40"),
    plot.caption  = element_text(size = 8,  color = "gray50", hjust = 1),
    panel.grid.major.x = element_blank(),
    panel.grid.minor   = element_blank(),
    axis.title    = element_text(size = 11),
    legend.position = "bottom"
  )


# ── Chart R1: Hub gene degrees (ggplot2 version) ──────────────────────────────

hub_data <- data.frame(
  gene   = c("CDKN1A", "JUN", "CCL2", "GDF15", "BTG2",
             "FOSL1",  "IER3", "CYCS", "HSPA1B", "S100A11"),
  degree = c(23, 18, 15, 10, 8, 8, 8, 8, 5, 4),
  type   = c("Ferroptosis", "Apoptosis", "Apoptosis", "Ferroptosis", "Apoptosis",
             "Apoptosis", "Apoptosis", "Apoptosis", "Apoptosis", "Apoptosis")
)
hub_data$gene <- factor(hub_data$gene, levels = hub_data$gene[order(hub_data$degree)])

p1 <- ggplot(hub_data, aes(x = degree, y = gene, fill = type)) +
  geom_col(width = 0.6) +
  geom_text(aes(label = degree), hjust = -0.3, size = 3.5, fontface = "bold") +
  scale_fill_manual(values = c("Ferroptosis" = CORAL, "Apoptosis" = TEAL),
                    name = "Gene type") +
  xlim(0, 28) +
  labs(
    title    = "Hub gene interaction degrees in co-expressed ferroptosis–apoptosis network",
    subtitle = "CDKN1A (ferroptosis) has the highest degree with 23 interactions",
    x        = "Number of interactions (degree)",
    y        = NULL,
    caption  = source_note
  ) +
  theme_thesis

ggsave("charts/R_chart1_hub_degrees.png", p1, width = 9, height = 5, dpi = 150)
cat("Saved: R_chart1_hub_degrees.png\n")


# ── Chart R2: KEGG pathway enrichment ─────────────────────────────────────────

kegg_data <- data.frame(
  pathway = c("Pathways in cancer", "Colorectal cancer", "Small cell lung cancer",
              "HTLV-1 infection", "Renal cell carcinoma",
              "Platinum drug resistance", "p53 signalling pathway"),
  genes   = c(4, 3, 3, 3, 2, 2, 2),
  highlight = c(FALSE, FALSE, FALSE, FALSE, FALSE, TRUE, TRUE)
)
kegg_data$pathway <- factor(kegg_data$pathway,
                             levels = kegg_data$pathway[order(kegg_data$genes)])

p2 <- ggplot(kegg_data, aes(x = genes, y = pathway, fill = highlight)) +
  geom_col(width = 0.55) +
  geom_text(aes(label = paste0(genes, " gene", ifelse(genes > 1, "s", ""))),
            hjust = -0.15, size = 3.5) +
  scale_fill_manual(values = c("FALSE" = PURPLE, "TRUE" = CORAL),
                    labels = c("Other cancer pathway", "Clinically significant to OC therapy"),
                    name = NULL) +
  xlim(0, 6.5) +
  labs(
    title    = "KEGG pathway enrichment of ferroptosis–apoptosis hub genes",
    subtitle = "Platinum drug resistance and p53 signalling are most relevant to ovarian cancer",
    x        = "Number of hub genes in pathway",
    y        = NULL,
    caption  = source_note
  ) +
  theme_thesis

ggsave("charts/R_chart2_kegg_pathways.png", p2, width = 10, height = 5, dpi = 150)
cat("Saved: R_chart2_kegg_pathways.png\n")


# ── Chart R3: DEG filtering funnel ────────────────────────────────────────────

funnel_data <- data.frame(
  stage = c("Total DEGs\n(with ref GTF)", "Significant\n(p < 0.05)",
            "Highly significant\n(p < 0.01)", "Ferroptosis +\nApoptosis genes"),
  count = c(7862, 8002, 6577, 289),
  fill  = c("start", "sig", "sig", "final")
)
funnel_data$stage <- factor(funnel_data$stage, levels = funnel_data$stage)

p3 <- ggplot(funnel_data, aes(x = stage, y = count, fill = fill)) +
  geom_col(width = 0.5) +
  geom_text(aes(label = format(count, big.mark = ",")),
            vjust = -0.5, fontface = "bold", size = 4) +
  scale_fill_manual(values = c("start" = "#D3D1C7", "sig" = TEAL, "final" = CORAL),
                    guide = "none") +
  scale_y_continuous(labels = scales::comma, expand = expansion(mult = c(0, 0.1))) +
  labs(
    title   = "Differential gene expression filtering funnel",
    subtitle = "From 7,862 total DEGs to 289 ferroptosis + apoptosis candidate genes",
    x       = NULL,
    y       = "Number of genes",
    caption = source_note
  ) +
  theme_thesis

ggsave("charts/R_chart3_deg_funnel.png", p3, width = 9, height = 5, dpi = 150)
cat("Saved: R_chart3_deg_funnel.png\n")


# ── Chart R4: Alignment scores ────────────────────────────────────────────────

align_data <- data.frame(
  sample  = c("SRR25637181", "SRR25637182", "SRR25637208*", "SRR25637216",
              "SRR25637217", "SRR25637220", "SRR25637224*",
              "SRR10313349", "SRR10313350", "SRR10313351",
              "SRR10313352", "SRR10313353", "SRR10313354"),
  score   = c(97.23, 97.56, 42.36, 96.88, 97.48, 97.86, 56.23,
              96.30, 96.29, 96.48, 96.26, 96.42, 96.54),
  group   = c(rep("Diseased", 7), rep("Control", 6)),
  passed  = c(TRUE, TRUE, FALSE, TRUE, TRUE, TRUE, FALSE,
              TRUE, TRUE, TRUE, TRUE, TRUE, TRUE)
)
align_data$sample <- factor(align_data$sample, levels = align_data$sample)

p4 <- ggplot(align_data, aes(x = sample, y = score, fill = interaction(group, passed))) +
  geom_col(width = 0.65) +
  geom_hline(yintercept = 80, linetype = "dashed", color = "red", alpha = 0.7) +
  annotate("text", x = 1, y = 82, label = "80% threshold", color = "red",
           size = 3, hjust = 0) +
  scale_fill_manual(
    values = c("Diseased.TRUE" = CORAL, "Diseased.FALSE" = "#F09595",
               "Control.TRUE"  = TEAL,  "Control.FALSE"  = "#9FE1CB"),
    labels = c("Diseased (passed)", "Diseased (excluded)",
               "Control (passed)",  "Control (excluded)"),
    name = NULL
  ) +
  scale_y_continuous(labels = function(x) paste0(x, "%"), limits = c(0, 105)) +
  labs(
    title   = "HISAT2 alignment scores — diseased vs control samples",
    subtitle = "Samples marked * were excluded (score < 80%)",
    x       = NULL,
    y       = "Alignment score (%)",
    caption = source_note
  ) +
  theme_thesis +
  theme(axis.text.x = element_text(angle = 40, hjust = 1, size = 8))

ggsave("charts/R_chart4_alignment_scores.png", p4, width = 12, height = 5, dpi = 150)
cat("Saved: R_chart4_alignment_scores.png\n")

cat("\nAll R charts generated in charts/ folder.\n")

File: thesis_visualizations.py
"""
Visualizations — Ferroptosis Hub Genes & Apoptosis in Ovarian Cancer
MSc Bioinformatics Dissertation, Pondicherry University (2024)
Author: Shruti Banerjee

Generates 5 charts from the thesis data using Python.
Run: python visualizations.py
Output: PNG files saved to charts/ folder.

Requirements: pip install matplotlib numpy pandas
"""

import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np
import os

os.makedirs("charts", exist_ok=True)

plt.rcParams.update({
    'font.family': 'sans-serif',
    'axes.spines.top': False,
    'axes.spines.right': False,
    'axes.grid': True,
    'grid.alpha': 0.25,
    'grid.linestyle': '--',
    'figure.dpi': 150
})

PURPLE = '#534AB7'
TEAL = '#1D9E75'
CORAL = '#D85A30'
AMBER = '#BA7517'
GRAY = '#888780'
LIGHT = '#D3D1C7'
SOURCE = 'Source: Banerjee S. (2024). MSc Dissertation, Pondicherry University.'


# ── Chart 1: QC Pipeline — sample count ───────────────────────────────────────

fig, ax = plt.subplots(figsize=(9, 5))

stages = ['Raw data\ncollected', 'After\nQC filter', 'After alignment\nscore filter']
diseased = [15, 7, 5]
control = [12, 6, 6]
x = np.arange(len(stages))
w = 0.35

b1 = ax.bar(x - w/2, diseased, w, label='Diseased samples', color=CORAL, zorder=3, borderradius=3)
b2 = ax.bar(x + w/2, control, w, label='Control samples', color=TEAL, zorder=3)

for bar in list(b1) + list(b2):
    ax.annotate(str(int(bar.get_height())),
                xy=(bar.get_x() + bar.get_width()/2, bar.get_height()),
                xytext=(0, 5), textcoords='offset points',
                ha='center', fontsize=11, fontweight='bold')

ax.set_xticks(x)
ax.set_xticklabels(stages)
ax.set_ylabel('Number of samples')
ax.set_ylim(0, 20)
ax.set_title('Sample count through the RNA-Seq QC and alignment pipeline\n(Ovarian cancer vs healthy controls)', fontsize=12, pad=12)
ax.legend()
ax.text(0.99, 0.02, SOURCE, transform=ax.transAxes, fontsize=8, color='gray', ha='right')
plt.tight_layout()
plt.savefig('charts/chart1_qc_pipeline.png', bbox_inches='tight')
print("Saved: chart1_qc_pipeline.png")
plt.close()


# ── Chart 2: HISAT2 alignment scores ──────────────────────────────────────────

fig, ax = plt.subplots(figsize=(10, 5))

d_ids = ['SRR25637181', 'SRR25637182', 'SRR25637208*', 'SRR25637216', 'SRR25637217', 'SRR25637220', 'SRR25637224*']
d_scores = [97.23, 97.56, 42.36, 96.88, 97.48, 97.86, 56.23]
c_ids = ['SRR10313349', 'SRR10313350', 'SRR10313351', 'SRR10313352', 'SRR10313353', 'SRR10313354']
c_scores = [96.30, 96.29, 96.48, 96.26, 96.42, 96.54]

x_d = np.arange(len(d_ids))
x_c = np.arange(len(c_ids))

fig, axes = plt.subplots(1, 2, figsize=(13, 5))

colors_d = [CORAL if s > 80 else '#F09595' for s in d_scores]
axes[0].bar(x_d, d_scores, color=colors_d, zorder=3)
axes[0].axhline(y=80, color='red', linestyle='--', alpha=0.6, label='Threshold (80%)')
axes[0].set_xticks(x_d)
axes[0].set_xticklabels(d_ids, rotation=35, ha='right', fontsize=9)
axes[0].set_ylim(0, 105)
axes[0].set_ylabel('Alignment score (%)')
axes[0].set_title('Diseased samples\n(* = excluded)', fontsize=11)
axes[0].legend(fontsize=9)
axes[0].grid(alpha=0.25, linestyle='--')
axes[0].spines['top'].set_visible(False)
axes[0].spines['right'].set_visible(False)
for i, v in enumerate(d_scores):
    axes[0].text(i, v + 1, f'{v}%', ha='center', fontsize=8)

axes[1].bar(x_c, c_scores, color=TEAL, zorder=3)
axes[1].axhline(y=80, color='red', linestyle='--', alpha=0.6, label='Threshold (80%)')
axes[1].set_xticks(x_c)
axes[1].set_xticklabels(c_ids, rotation=35, ha='right', fontsize=9)
axes[1].set_ylim(0, 105)
axes[1].set_ylabel('Alignment score (%)')
axes[1].set_title('Control samples\n(all passed QC)', fontsize=11)
axes[1].legend(fontsize=9)
axes[1].grid(alpha=0.25, linestyle='--')
axes[1].spines['top'].set_visible(False)
axes[1].spines['right'].set_visible(False)
for i, v in enumerate(c_scores):
    axes[1].text(i, v + 1, f'{v}%', ha='center', fontsize=8)

fig.suptitle('HISAT2 alignment scores — diseased vs control samples\n(Reference genome: ENSEMBL GRCh38)', fontsize=12, y=1.02)
fig.text(0.99, -0.06, SOURCE, fontsize=8, color='gray', ha='right')
plt.tight_layout()
plt.savefig('charts/chart2_alignment_scores.png', bbox_inches='tight')
print("Saved: chart2_alignment_scores.png")
plt.close()


# ── Chart 3: Hub gene interaction degrees ─────────────────────────────────────

fig, ax = plt.subplots(figsize=(9, 5))

genes = ['CDKN1A\n(Ferroptosis)', 'JUN\n(Apoptosis)', 'CCL2\n(Apoptosis)',
         'GDF15\n(Ferroptosis)', 'BTG2\n(Apoptosis)', 'FOSL1\n(Apoptosis)',
         'IER3\n(Apoptosis)', 'CYCS\n(Apoptosis)', 'HSPA1B\n(Apoptosis)', 'S100A11\n(Apoptosis)']
degrees = [23, 18, 15, 10, 8, 8, 8, 8, 5, 4]
colors = [CORAL if 'Ferroptosis' in g else TEAL for g in genes]

bars = ax.barh(genes, degrees, color=colors, zorder=3, height=0.6)

for bar, val in zip(bars, degrees):
    ax.text(bar.get_width() + 0.3, bar.get_y() + bar.get_height()/2,
            str(val), va='center', fontsize=10, fontweight='bold')

ax.set_xlabel('Number of interactions (degree)')
ax.set_title('Hub gene interaction degrees in co-expressed\nferroptosis–apoptosis network', fontsize=12, pad=12)
ax.set_xlim(0, 28)

legend_items = [
    mpatches.Patch(color=CORAL, label='Ferroptosis gene'),
    mpatches.Patch(color=TEAL, label='Apoptosis gene')
]
ax.legend(handles=legend_items, loc='lower right')
ax.text(0.99, 0.02, SOURCE, transform=ax.transAxes, fontsize=8, color='gray', ha='right')
plt.tight_layout()
plt.savefig('charts/chart3_hub_gene_degrees.png', bbox_inches='tight')
print("Saved: chart3_hub_gene_degrees.png")
plt.close()


# ── Chart 4: KEGG pathway gene involvement ────────────────────────────────────

fig, ax = plt.subplots(figsize=(10, 5))

pathways = [
    'Pathways in cancer',
    'Colorectal cancer',
    'Small cell lung cancer',
    'Human T-cell\nleukemia virus 1',
    'Renal cell carcinoma',
    'Platinum drug resistance',
    'p53 signalling pathway'
]
gene_counts = [4, 3, 3, 3, 2, 2, 2]
highlight = ['Platinum drug resistance', 'p53 signalling pathway']
colors_k = [CORAL if p.strip() in highlight else PURPLE for p in pathways]

bars = ax.barh(pathways, gene_counts, color=colors_k, zorder=3, height=0.55)

for bar, val in zip(bars, gene_counts):
    ax.text(bar.get_width() + 0.05, bar.get_y() + bar.get_height()/2,
            f'{val} gene{"s" if val > 1 else ""}', va='center', fontsize=10)

ax.set_xlabel('Number of hub genes in pathway')
ax.set_title('KEGG pathway enrichment of co-expressed ferroptosis–apoptosis hub genes\n(highlighted = clinically significant to ovarian cancer therapy)', fontsize=11, pad=12)
ax.set_xlim(0, 6)

legend_items = [
    mpatches.Patch(color=CORAL, label='Clinically significant pathway'),
    mpatches.Patch(color=PURPLE, label='Other cancer pathway')
]
ax.legend(handles=legend_items, loc='lower right')
ax.text(0.99, 0.02, SOURCE, transform=ax.transAxes, fontsize=8, color='gray', ha='right')
plt.tight_layout()
plt.savefig('charts/chart4_kegg_pathways.png', bbox_inches='tight')
print("Saved: chart4_kegg_pathways.png")
plt.close()


# ── Chart 5: DEG filtering funnel ─────────────────────────────────────────────

fig, ax = plt.subplots(figsize=(8, 5))

stages_f = ['Total DEGs\n(with ref GTF)', 'Significant\n(p < 0.05)', 'Highly significant\n(p < 0.01)', 'Ferroptosis +\nApoptosis genes']
counts = [7862, 8002, 6577, 289]
colors_f = [LIGHT, TEAL, TEAL, CORAL]

bars = ax.bar(stages_f, counts, color=colors_f, zorder=3, width=0.5)

for bar, val in zip(bars, counts):
    ax.annotate(f'{val:,}',
                xy=(bar.get_x() + bar.get_width()/2, bar.get_height()),
                xytext=(0, 6), textcoords='offset points',
                ha='center', fontsize=11, fontweight='bold')

ax.set_ylabel('Number of genes')
ax.set_title('Differential gene expression filtering funnel\n(from raw count to ferroptosis + apoptosis candidate genes)', fontsize=12, pad=12)
ax.text(0.99, 0.02, SOURCE, transform=ax.transAxes, fontsize=8, color='gray', ha='right')
plt.tight_layout()
plt.savefig('charts/chart5_deg_funnel.png', bbox_inches='tight')
print("Saved: chart5_deg_funnel.png")
plt.close()

print("\nAll 5 charts generated in charts/ folder.")
